substrate/frame/asset-conversion/src/lib.rs

```diff
@@ -845,63 +957,85 @@ pub mod pallet {
-		pub(crate) fn do_swap(
-			sender: T::AccountId,
-			amounts: &Vec<T::AssetBalance>,
-			path: BoundedVec<T::MultiAssetId, T::MaxSwapPathLength>,
-			send_to: T::AccountId,
+		/// Swap assets along the `path`, withdrawing from `sender` and depositing in `send_to`.
+		///
+		/// Note: It's assumed that the provided `path` is valid.
+		///
+		/// WARNING: This may return an error after a partial storage mutation. It should be used
+		/// only inside a transactional storage context and an Err result must imply a storage
+		/// rollback.
+		fn swap(
+			sender: &T::AccountId,
+			path: &BalancePath<T>,
+			send_to: &T::AccountId,
 			keep_alive: bool,
 		) -> Result<(), DispatchError> {
-			ensure!(amounts.len() > 1, Error::<T>::CorrespondenceError);
-			if let Some([asset1, asset2]) = &path.get(0..2) {
-				let pool_id = Self::get_pool_id(asset1.clone(), asset2.clone());
-				let pool_account = Self::get_pool_account(&pool_id);
-				// amounts should always contain a corresponding element to path.
-				let first_amount = amounts.first().ok_or(Error::<T>::CorrespondenceError)?;
-
-				Self::transfer(asset1, &sender, &pool_account, *first_amount, keep_alive)?;
-
-				let mut i = 0;
-				let path_len = path.len() as u32;
-				for assets_pair in path.windows(2) {
-					if let [asset1, asset2] = assets_pair {
-						let pool_id = Self::get_pool_id(asset1.clone(), asset2.clone());
-						let pool_account = Self::get_pool_account(&pool_id);
-
-						let amount_out =
-							amounts.get((i + 1) as usize).ok_or(Error::<T>::CorrespondenceError)?;
-
-						let to = if i < path_len - 2 {
-							let asset3 = path.get((i + 2) as usize).ok_or(Error::<T>::PathError)?;
-							Self::get_pool_account(&Self::get_pool_id(
-								asset2.clone(),
-								asset3.clone(),
-							))
-						} else {
-							send_to.clone()
-						};
+			let (asset_in, amount_in) = path.first().ok_or(Error::<T>::InvalidPath)?;
+			let credit_in = Self::withdraw(asset_in, sender, *amount_in, keep_alive)?;
+
+			let credit_out = Self::credit_swap(credit_in, path).map_err(|(_, e)| e)?;
+			Self::resolve(send_to, credit_out).map_err(|_| Error::<T>::BelowMinimum)?;
+
+			Ok(())
+		}
 
-						let reserve = Self::get_balance(&pool_account, asset2)?;
+		/// Swap assets along the specified `path`, consuming `credit_in` and producing
+		/// `credit_out`.
+		///
+		/// If an error occurs, `credit_in` is returned back.
+		///
+		/// Note: It's assumed that the provided `path` is valid and `credit_in` corresponds to the
+		/// first asset in the `path`.
+		///
+		/// WARNING: This may return an error after a partial storage mutation. It should be used
+		/// only inside a transactional storage context and an Err result must imply a storage
+		/// rollback.
+		fn credit_swap(
+			credit_in: Credit<T>,
+			path: &BalancePath<T>,
+		) -> Result<Credit<T>, (Credit<T>, DispatchError)> {
+			let resolve_path = || -> Result<Credit<T>, DispatchError> {
```

Comment:
helps to avoid matching a non clone credit object for returning it in case of error 
alternative solution could be - https://gist.github.com/muharem/24376bcaa4ab19c3e7e0974cbc063648